### PR TITLE
OpenCensus: Use new CallTracer interfaces

### PR DESCRIPTION
**Pull request by yashykt**
This change mostly aims to get OpenCensus to use the new ServerCallTracer interface. Note that the interfaces nor the code are in their final states. There are a bunch of moving pieces, but I thought this might be a nice mid-step to check-in and make sure that our internal traces can also work with these changes.

Overall changes -
1) call_tracer.h shows what the hierarchy of new call tracer interfaces looks like. Open to renaming suggestions.
2) Moved most of the common interface between `CallAttemptTracer` and `ServerCallTracer` into a common `CallTracerInterface`. We should be able to eventually move `RecordReceivedTrailingMetadata` and `RecordEnd` as well to these common interfaces, but it requires some additional work.
3) The compression filter is now responsible for recording the recv and send messages for both the subchannel call and the server, and adds in ability to record compressed and decompressed messages as well.
4) The OpenCensus server filter now uses the new `ServerCallTracer` interface, and so doesn't need to be a filter anymore.
5) A new ServerCallTracerFilter was added. Ideally, we should be able to move it to the current connected filter, but it is in a bit of an interesting state right now, so I would prefer making those changes in a separate PR with Craig's eyes on it.
6) A new context element `GRPC_CONTEXT_CALL_TRACER_ANNOTATION_INTERFACE` was created that replaces the old `GRPC_CONTEXT_CALL_TRACER`, and the new `GRPC_CONTEXT_CALL_TRACER` is mainly to pass the `CallAttemptTracer` down the stack. This should go away in the new promise-based world.

- - -
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/32618)

### CODE REVIEW

File: src/core/plugin_registry/grpc_plugin_registry.cc
Comment (markdroth) [inline src/core/plugin_registry/grpc_plugin_registry.cc:113]: Is this positioning ensuring that this filter will be the last one in the stack, right before the connected_channel filter?

File: src/cpp/ext/filters/census/grpc_plugin.cc
Comment (markdroth) [inline src/cpp/ext/filters/census/grpc_plugin.cc:54]: Why do we no longer need the server side to be a filter but we still need the client side to be a filter?

File: src/cpp/ext/filters/census/server_filter.h
Comment (markdroth) [inline src/cpp/ext/filters/census/server_filter.h:1]: This file should be renamed, since it's no longer a filter.

Same for the .cc file.

File: test/core/xds/xds_channel_stack_modifier_test.cc
Comment (markdroth) [inline test/core/xds/xds_channel_stack_modifier_test.cc:114]: Shouldn't we still test this with "census_server", since the code is still checking for that case?

**Review by markdroth — COMMENTED** [review]
This looks like a great start!

Please let me know if you have any questions.  Thanks!

**Review by yashykt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/32618)* status: 0 of 33 files reviewed, 7 unresolved discussions (waiting on @ctiller, @markdroth, and @veblush)

___
*[`src/core/lib/channel/call_tracer.h` line 60 at r1](https://reviewable.io/reviews/grpc/grpc/32618#-NQaT_aUz5XJysFB59ry-r1-60:-NQaUNoF1cmwiQF4JWeB:b8do4i0) ([raw file](https://github.com/grpc/grpc/blob/3461c11972b7b8f7eccf2cc8736ffc8835e71160/src/core/lib/channel/call_tracer.h#L60)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I am raising the following idea just to make sure that we've considered alternatives -- my guess is that we don't actually want to do this.

Since we're now going to be using this same interface on both the client and server side, would it make sense to phrase these methods in terms of client and server rather than send and recv?  For example, we could have a method like `RecordClientInitialMetadata()`, which would be called for send_initial_metadata on the client side and recv_initial_metadata on the server side.

The motivation for this idea is that a lot of the promise code is phrased this way, so I'm wondering if this approach would be more consistent in the long run.

When evaluating this idea, please consider both (a) how the tracer implementations will work and (b) how the `RpcTracerInterface` API will be called in various parts of the code (e.g., filters, transport, etc).  The overall goal here should be minimizing `if (is_client)` throughout the code base.  (Looking at the compression filter, it looks like the send/recv pattern is actually better than the client/server pattern would be.  But I don't know if that's also the case for the other call sites that we'll have in the future.)

One other factor to consider is that I think there are cases where the mutability of the parameters will want to be different on the client and server side.  For example, client initial metadata needs to be mutable on the client side but can be const on the server side, whereas server trailing(?) metadata needs to be mutable on the server side but can be const on the client side.  If we use the same interface for both sides, then we need to make the parameter mutable even on the side where we don't really want to support mutation.  However, if we stick with the current send/recv interface, then I think we can make the parameters mutable only for the send operations.

I'd like to get @ctiller's input on this as well.
</blockquote></details>

I did consider this, and had similar considerations as you - 
1) For example for compression/decompression, we only need to care about whether we are receiving and need to decompress, or whether we are sending and need to decompress.
2) As you suggest, the mutability of parameters is also a good point. We wouldn't need to modify metadata on the receive side, just on the send side.
___
*[`src/core/lib/channel/call_tracer.h` line 95 at r1](https://reviewable.io/reviews/grpc/grpc/32618#-NQaT_aUz5XJysFB59ry-r1-95:-NQaVC4B7DNvosW9KAOC:bfxhcdk) ([raw file](https://github.com/grpc/grpc/blob/3461c11972b7b8f7eccf2cc8736ffc8835e71160/src/core/lib/channel/call_tracer.h#L95)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

What additional work is required to move these last two methods to `RpcTracerInterface`?
</blockquote></details>

The first change would be to split the final status and stats from `RecordReceivedTrailingMetadata` and move it to a separate call (maybe potentially move it to `RecordEnd`). `RecordEnd` would also need to change accordingly. This part seems potentially tricky enough that I would want a dedicated PR that has reviewers just look at that change.
___
*[`src/core/lib/channel/context.h` line 45 at r1](https://reviewable.io/reviews/grpc/grpc/32618#-NQaT_aUz5XJysFB59rz-r1-45:-NQaXjAuCs73J9GFC_ZZ:b-sh1mqz) ([raw file](https://github.com/grpc/grpc/blob/3461c11972b7b8f7eccf2cc8736ffc8835e71160/src/core/lib/channel/context.h#L45)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Suggest renaming `GRPC_CONTEXT_CALL_TRACER` to `GRPC_CONTEXT_CALL_ATTRIBUTE_TRACER` and then using `GRPC_CONTEXT_CALL_TRACER` for this new one.
</blockquote></details>

What does `GRPC_CONTEXT_CALL_ATTRIBUTE_TRACER` even mean? :)
This is messy since with these changes GRPC_CONTEXT_CALL_ATTRIBUTE_TRACER would hold a  client-side`CallTracer`, while a `GRPC_CONTEXT_CALL_TRACER` would now hold a `CallAttemptTracer`
___
*[`src/core/lib/surface/call.cc` line 749 at r1](https://reviewable.io/reviews/grpc/grpc/32618#-NQaT_aUz5XJysFB59s0-r1-749:-NQaYpbl74Mh5BNCDfIh:b-jzyr5x) ([raw file](https://github.com/grpc/grpc/blob/3461c11972b7b8f7eccf2cc8736ffc8835e71160/src/core/lib/surface/call.cc#L749)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I think we decided that we want the tracer to be created in the AcceptStream callback, right?  Can't we do it there immediately?
</blockquote></details>

No, since we need the arena to create the `ServerCallTracer`, we said we'll have AcceptStream return the `ServerCallTracer` object to the transport (not yet done).
___
*[`src/core/plugin_registry/grpc_plugin_registry.cc` line 113 at r1](https://reviewable.io/reviews/grpc/grpc/32618#-NQaT_aUz5XJysFB59s1-r1-113:-NQaZ6D29IzzJjjh0gbx:bnp90j3) ([raw file](https://github.com/grpc/grpc/blob/3461c11972b7b8f7eccf2cc8736ffc8835e71160/src/core/plugin_registry/grpc_plugin_registry.cc#L113)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Is this positioning ensuring that this filter will be the last one in the stack, right before the connected_channel filter?
</blockquote></details>

That's a good point. Hmm, I got a smart idea, I can actually register it along with the "connected" filter for now as a built_in filter.

I still want to avoid having this filter and move it to the connected filter, but the connected filter seems to be in a messy state right now, so this might still be the right thing to do for now.

**Review by yashykt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/32618)* status: 0 of 33 files reviewed, 7 unresolved discussions (waiting on @ctiller, @markdroth, and @veblush)

___
*[`src/core/ext/filters/client_channel/client_channel.cc` line 2529 at r1](https://reviewable.io/reviews/grpc/grpc/32618#-NQaT_aUz5XJysFB59rs-r1-2529:-NQagL-G-wkmz-vGI0Nr:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/3461c11972b7b8f7eccf2cc8736ffc8835e71160/src/core/ext/filters/client_channel/client_channel.cc#L2529)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Now that we're storing this in the call context, I don't think we need this data member anymore.  Instead, we can get it from `call_context()[GRPC_CONTEXT_RPC_TRACER]`.
</blockquote></details>

Done.
___
*[`src/core/lib/channel/call_tracer.h` line 54 at r1](https://reviewable.io/reviews/grpc/grpc/32618#-NQaT_aUz5XJysFB59ry-r1-54:-NQagLlB1Uhkv_XDhKkg:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/3461c11972b7b8f7eccf2cc8736ffc8835e71160/src/core/lib/channel/call_tracer.h#L54)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I suggest renaming `CallTracerInterface` to something like `CallTracerAnnotationInterface` and then renaming `RpcTracerInterface` to `CallTracerInterface`.
</blockquote></details>

Done.

File: src/core/ext/filters/client_channel/client_channel.cc
Comment (yashykt) [reply src/core/ext/filters/client_channel/client_channel.cc:3055]: Thanks! Removed!

File: src/core/lib/channel/call_tracer.h
Comment (yashykt) [reply src/core/lib/channel/call_tracer.h:58]: (please view discussion on reviewable)

File: src/core/lib/channel/call_tracer.h
Comment (yashykt) [reply src/core/lib/channel/call_tracer.h:64]: (please view discussion on reviewable)

File: src/core/lib/channel/call_tracer.h
Comment (yashykt) [reply src/core/lib/channel/call_tracer.h:99]: (please view discussion on reviewable)

File: src/core/lib/channel/context.h
Comment (yashykt) [reply src/core/lib/channel/context.h:46]: Done

File: src/core/lib/surface/call.cc
Comment (yashykt) [reply src/core/lib/surface/call.cc:748]: (please view discussion on reviewable)

File: src/cpp/ext/filters/census/grpc_plugin.cc
Comment (yashykt) [reply src/cpp/ext/filters/census/grpc_plugin.cc:54]: As discussed in the design review, the method of creating call tracers are going to be different on the client-side and the server-side.
On the client-side, we are going to have an interceptor inject this. (I know it's a filter right now, but it'll move to an interceptor). On the server-side, we will have a factory that gets invoked by the surface call.

File: test/core/xds/xds_channel_stack_modifier_test.cc
Comment (yashykt) [reply test/core/xds/xds_channel_stack_modifier_test.cc:114]: This test just checks for OpenCensus unfortunately :( For the internal census, we will need a test in google3.

File: src/core/lib/channel/server_call_tracer.h
Comment (yashykt) [reply src/core/lib/channel/server_call_tracer.h:1]: Done

File: src/core/plugin_registry/grpc_plugin_registry.cc
Comment (yashykt) [reply src/core/plugin_registry/grpc_plugin_registry.cc]: (please view discussion on reviewable)

File: src/cpp/ext/filters/census/server_filter.h
Comment (yashykt) [reply src/cpp/ext/filters/census/server_filter.h:1]: Done

File: src/core/ext/filters/client_channel/client_channel.cc
Comment (yashykt) [reply src/core/ext/filters/client_channel/client_channel.cc]: Done.

File: test/core/xds/xds_channel_stack_modifier_test.cc
Comment (yashykt) [reply test/core/xds/xds_channel_stack_modifier_test.cc:114]: cl/516905576 moves it to an internal test

File: src/core/lib/channel/call_tracer.h
Comment (markdroth) [inline src/core/lib/channel/call_tracer.h:87]: Suggest renaming this to `ClientCallTracer` for clarity.

File: src/core/lib/channel/context.h
Comment (markdroth) [inline src/core/lib/channel/context.h:41]: This can move up to the previous line.

File: src/core/lib/channel/call_tracer.h
Comment (markdroth) [reply src/core/lib/channel/call_tracer.h:99]: Suggest adding a TODO about this.

**Review by markdroth — APPROVED** [review]
This LGTM, modulo a few remaining comments.

I'd also like Craig to weigh in on the design question, but that does not have to block this PR.

File: src/core/lib/channel/call_tracer.h
Comment (yashykt) [reply src/core/lib/channel/call_tracer.h]: Done

File: src/core/lib/channel/context.h
Comment (yashykt) [reply src/core/lib/channel/context.h]: Done

**Comment by ctiller** [comment]
If you're waiting for my input, please feel free to ping me on chat.
It'd be especially nice to get a heads up on reviews on promise based calls right now.

File: src/core/lib/channel/call_tracer.h
Comment (devbww) [inline src/core/lib/channel/call_tracer.h:44]: A backslash before a newline is probably something you want to avoid.

**Pull request merged**
